### DISCUSSION TITLE
Using custom `com.microsoft` ops from python

### DISCUSSION THREAD

**Discussion by brnorris03** [microsoft/onnxruntime#22023] [API Q&A]
I would like to use the "com.microsoft" domain ops whose schemas are available in `onnxruntime`, e.g., ["SoftmaxGrad"](https://discord.com/channels/@me/1176668637369749576/1281709195988570113), but cannot figure out how after looking at docs and the codebase. I am configuring with `--enable_training --enable_pybind --enable_external_custom_op_schemas`. The `onnxruntime.training` python package builds successfully. 

I am new to this, apologies for the possibly very basic questions. Here is a simple test I tried first:
```python
from onnx import checker, helper
opset_import = [
    helper.make_opsetid(domain="com.microsoft", version=1),
    helper.make_opsetid(domain="", version=17),
]

node = helper.make_node(
    op_type="SoftmaxGrad",
    name="/SoftmaxGrad_0",
    inputs=["dY", "Y"],
    outputs=["dX"],
    axis=-1,
    domain="com.microsoft",
    opset_imports=opset_import,
)
checker.check_node(node)
```
which fails with
```
Traceback (most recent call last):
  File "/proj/work/bnorris/mlir-compiler/ex.py", line 10, in <module>
    node = helper.make_node(
  File "/.../venv/lib/python3.10/site-packages/onnx/helper.py", line 173, in make_node
    node.attribute.extend(
  File "/.../venv/lib/python3.10/site-packages/onnx/helper.py", line 174, in <genexpr>
    make_attribute(key, value)
  File "/.../venv/lib/python3.10/site-packages/onnx/helper.py", line 960, in make_attribute
    raise ValueError(
ValueError: Could not infer the attribute type from the elements of the passed Iterable value.
```

I did check `onnx.defs.get_all_schemas_with_history()`, which does not include the custom ops.

I also attempted to use the custom op documentation generator, which relies on the `schemadef` package, but that did not appear to have been compiled into the python bindings. It seems to me that none of the python functions in `onnxruntime/python/onnxruntime_pybind_schema.cc` are being compiled, and I can't figure out what controls whether they are built or not.

The problem (taken from the documentation generation script):

```python
from onnxruntime.capi.onnxruntime_pybind11_state import schemadef

ImportError: cannot import name 'schemadef' from 'onnxruntime.capi.onnxruntime_pybind11_state' (/.../venv/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_pybind11_state.so)
```

**Comment by xadupre** [comment]
Your model is invalid. You need to create a full model with make_model, make_graph, make_node... See https://onnx.ai/onnx/intro/python.html.

**Comment by brnorris03** [comment]
I don't have a model for this test, just a single node (but creating a model does not work since it fails at node creation). Can you elaborate why exactly I need a model to invoke `check_node`? I don't see this requirement mentioned anywhere in code or documentation.

The above test is based on https://github.com/onnx/onnx/blob/67c456ba4747412afb44158a1a889c0fc3349641/onnx/test/checker_test.py, which suggests I should be able to create a node and check it without inserting it into a full model. 

Thanks!

**Reply by Craigacp** [reply]
Isn't the opset_imports a property of the model rather than a property of a node? So it's probably trying to convert the `opset_imports` argument into an ONNX attribute and failing.
